xDiffCollection/CollectionBin.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  CollectionBin.swift
+//  CollectionTest
+//
+//  Created by Esteban Garro on 2018-06-05.
+//  Copyright © 2018 aldo. All rights reserved.
+//
+
+import Foundation
+
+internal enum BinOperationType {
+    case updated
+    case added
+    case deleted
+    case none
+}
+
+internal struct BinResult {
+    var type: BinOperationType
+    var idx : Int
+}
+
+internal protocol CollectionFiltering {
+    func filter<T:Hashable>(_ element:T) -> Bool
+}
+
+internal struct CollectionBin<T:Hashable> : CollectionFiltering, CustomStringConvertible {
+    var index : Int
+    var elements:Array<T> = []
+    var filter:(T)->Bool
+    var description: String {
+        var resp = ""
+        for (idx, element) in self.elements.enumerated() {
+            resp += "\(idx))\t\(element.hashValue)\n"
+        }
+        return resp
+    }
+    var count : Int {
+        return elements.count
+    }
+    
+    init(index: Int, filter: @escaping (T)->Bool, elements: Array<T>) {
+        self.index = index
+        self.filter = filter
+        self.elements = elements
+    }
+    
+    func filter<T>(_ element:T) -> Bool {
+        return self.filter(element)
+    }
+    
+    func updated(_ element:T) -> (CollectionBin<T>, BinResult) {
+        var copy = elements
+        if let idx = self.contains(element) {
+            copy[idx] = element
+            return (CollectionBin(index: self.index, filter: self.filter, elements: copy),BinResult(type: .updated, idx: idx))
+        } else {
+            copy.append(element)
+            return (CollectionBin(index: self.index, filter: self.filter, elements: copy),BinResult(type: .added, idx: copy.count-1))
+        }
+    }
+    
+    func removed(_ element:T) -> (CollectionBin<T>, BinResult) {
+        var copy = elements
+        if let idx = self.contains(element) {
```

Comment:
Want to try with filter? 
Looks like we could've conformed to collection protocol

Reply:
Same here! In fact now that I think about it, let's look for parts of the code that are not written as per our functional-programming standards, and we'll use them as exercise that we can do together! These are very good examples.

Reply:
Same reply as above

Reply:
Same response as above 😛

Reply:
agree again!